src/OrbitHttp/HttpDownloadManagerTest.cpp

```diff
@@ -0,0 +1,210 @@
+
+    QProcessEnvironment current_env = local_http_server_process_->processEnvironment();
+    current_env.insert("PYTHONUNBUFFERED", "true");
+    local_http_server_process_->setProcessEnvironment(current_env);
+
+    ORBIT_LOG("Execute command:\n\"%s %s\"\n", local_http_server_process_->program().toStdString(),
+              local_http_server_process_->arguments().join(" ").toStdString());
+
+    QEventLoop loop{};
+    QObject::connect(local_http_server_process_, &QProcess::readyReadStandardOutput,
+                     [&loop, this]() {
+                       const QString prefix = "Serving HTTP on";
+                       QString std_output = local_http_server_process_->readAllStandardOutput();
+                       if (!std_output.contains(prefix)) return;
+
+                       QRegularExpression portRegex("port ([0-9]+)");
+                       QRegularExpressionMatch portMatch = portRegex.match(std_output);
+                       if (portMatch.hasMatch()) {
+                         port_ = portMatch.captured(1);
+                         loop.quit();
+                       }
+                     });
+
+    QObject::connect(local_http_server_process_, &QProcess::errorOccurred,
+                     [&loop, this](QProcess::ProcessError error) {
+                       if (error == QProcess::Crashed) return;
+                       ORBIT_LOG("Error while executing process.\nError:\n%s,\nDetails:\n%s.\n",
+                                 QMetaEnum::fromType<QProcess::ProcessError>().valueToKey(error),
+                                 local_http_server_process_->errorString().toStdString());
+                       if (loop.isRunning()) loop.quit();
+                     });
+
+    local_http_server_process_->start();
+    loop.exec();
+  }
+
+  ~HttpDownloadManagerTest() override {
+    local_http_server_process_->kill();
+    for (const auto& file_path : files_to_remove_) (void)orbit_base::RemoveFile(file_path);
+  }
+
+  [[nodiscard]] std::filesystem::path GetTemporaryFilePath() {
+    auto temporary_file_or_error = orbit_base::TemporaryFile::Create();
+    EXPECT_THAT(temporary_file_or_error, HasNoError());
+
+    TemporaryFile temporary_file = std::move(temporary_file_or_error.value());
+    std::filesystem::path file_path = temporary_file.file_path();
+    temporary_file.CloseAndRemove();
+
+    // No matter the file is downloaded or not, we will try to remove it in the end.
+    files_to_remove_.push_back(file_path);
+
+    return file_path;
+  }
+
+  [[nodiscard]] std::string GetUrl(std::string filename) const {
+    return absl::StrFormat("http://localhost:%s/%s", port_.toStdString(), filename);
+  }
+
+  std::shared_ptr<orbit_qt_utils::MainThreadExecutorImpl> executor_;
+  HttpDownloadManager manager_;
+
+ private:
+  QProcess* local_http_server_process_;
+  std::vector<std::filesystem::path> files_to_remove_;
+  QString port_;
+};
+}  // namespace
+
+TEST_F(HttpDownloadManagerTest, DownloadSingleSucceeded) {
```

Comment:
Can you also add a test case where you destruct the download manager while a download is in progress, to ensure that nothing crashes and the download is properly aborted?

Reply:
And can you also add a test case that tests the "Cannot write to file" case? I would assume if you were providing a `local_path` where the parent directory does not exist, then it should fail to open the file and the download should also fail.

Reply:
I have added the DownloadSingleInvalidSaveFilePath and DownloadSingleDestroyManagerEarly to the test.

Sorry that I have to bring the HttpDownloadOperation back to the public include directory...